controllers/add_geteway.go

```diff
@@ -0,0 +1,176 @@
+			return &requeue{curError: err}
+		}
+		if err = ctrl.SetControllerReference(hdb, &deploy, r.Scheme); err != nil {
+			return &requeue{curError: err}
+		}
+
+		logger.Info("Create gateway")
+		if err = r.Client.Create(ctx, &deploy); err != nil {
+			return &requeue{curError: err}
+		}
+		return nil
+	}
+	if !isHashChanged(&existingDeploy.ObjectMeta, &deploy.ObjectMeta) {
+		return nil
+	}
+
+	logger.Info("Update gateway")
+	r.Recorder.Event(hdb, corev1.EventTypeNormal, "UpdatingGateway", "")
+
+	existingDeploy.Annotations = deploy.Annotations
+	existingDeploy.Labels = deploy.Labels
+	existingDeploy.Spec = deploy.Spec
+	if err = r.Update(ctx, existingDeploy); err != nil {
+		return &requeue{curError: err}
+	}
+	return nil
+}
+
+func (a addGateway) getDeployment(ctx context.Context, r *HStreamDBReconciler, hdb *hapi.HStreamDB) appsv1.Deployment {
+	podTemplate := a.getPodTemplate(ctx, r, hdb)
+	deploy := internal.GetDeployment(hdb, &hdb.Spec.Gateway.Component,
+		&podTemplate, hapi.ComponentTypeGateway)
+
+	return deploy
+}
+
+func (a addGateway) getPodTemplate(ctx context.Context, r *HStreamDBReconciler, hdb *hapi.HStreamDB) corev1.PodTemplateSpec {
+	gateway := hdb.Spec.Gateway
+
+	pod := corev1.PodTemplateSpec{
+		ObjectMeta: internal.GetObjectMetadata(hdb, nil, hapi.ComponentTypeGateway),
+		Spec: corev1.PodSpec{
+			Affinity:        gateway.Affinity,
+			Tolerations:     gateway.Tolerations,
+			NodeName:        gateway.NodeName,
+			NodeSelector:    gateway.NodeSelector,
+			SecurityContext: gateway.PodSecurityContext,
+			InitContainers:  gateway.InitContainers,
+			Containers:      a.getContainer(ctx, r, hdb),
+			Volumes:         gateway.Volumes,
+		},
+	}
+	if gateway.SecretRef != nil {
+		pod.Spec.Volumes = append(pod.Spec.Volumes, corev1.Volume{
+			Name: "cert",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: gateway.SecretRef.Name,
+				},
+			},
+		})
+	}
+
+	pod.Name = hapi.ComponentTypeAdminServer.GetResName(hdb.Name)
+	return pod
+}
+
+func (a addGateway) getContainer(ctx context.Context, r *HStreamDBReconciler, hdb *hapi.HStreamDB) []corev1.Container {
+	gateway := hdb.Spec.Gateway
+	container := corev1.Container{
+		Image:           hdb.Spec.Gateway.Image,
+		ImagePullPolicy: hdb.Spec.Gateway.ImagePullPolicy,
+	}
+
+	structAssign(&container, gateway.Container.DeepCopy())
+
+	if container.Name == "" {
+		container.Name = string(hapi.ComponentTypeGateway)
+	}
+
+	container.Ports = extendPorts(map[string]string{},
+		container.Ports,
+		[]corev1.ContainerPort{
+			{Name: "port", ContainerPort: gateway.Port},
+		},
+	)
+
+	podList := corev1.PodList{}
+	_ = r.List(ctx, &podList,
+		client.InNamespace(hdb.Namespace),
+		client.MatchingLabels(
+			internal.GetObjectMetadata(hdb, nil, hapi.ComponentTypeHServer).Labels,
+		),
+	)
+
+	port := findHServerPort(hdb, podList.Items[0])
+	hServerSvc := internal.GetHeadlessService(hdb, hapi.ComponentTypeHServer, corev1.ServicePort{})
```

Comment:
The third param of func `internal.GetHeadlessService` is a variable parameter,  we only construct the svc name by calling this func, and no need to pass svc port, so just remove the param `corev1.ServicePort`, like this `internal.GetHeadlessService(hdb, hapi.ComponentTypeHServer)`